test/jdk/java/nio/channels/unixdomain/IOExchanges.java

```diff
@@ -0,0 +1,1237 @@
+import static java.nio.channels.SelectionKey.OP_ACCEPT;
+import static java.nio.channels.SelectionKey.OP_READ;
+import static java.nio.channels.SelectionKey.OP_WRITE;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertTrue;
+
+public class IOExchanges {
+    static boolean supported = true;
+
+
+    @BeforeTest()
+    public void setup() {
+        try {
+            SocketChannel.open(UNIX);
+        } catch (IOException | UnsupportedOperationException e) {
+            supported = false;
+            out.println("Unix domain channels not supported");
+        }
+    }
+
+    static SocketChannel openSocketChannel(ProtocolFamily family)
+            throws IOException {
+        return family == UNIX ? SocketChannel.open(family)
+                : SocketChannel.open();
+    }
+
+    static ServerSocketChannel openServerSocketChannel(ProtocolFamily family)
+             throws IOException {
+        return family == UNIX ? ServerSocketChannel.open(family)
+                     : ServerSocketChannel.open();
+    }
+
+    public static void deleteFile(SocketAddress addr) throws Exception {
+        if (addr instanceof UnixDomainSocketAddress) {
+            Files.deleteIfExists(((UnixDomainSocketAddress) addr).getPath());
+        }
+    }
+
+    /*
+     The following, non-exhaustive set, of tests exercise different combinations
+     of blocking and non-blocking accept/connect calls along with I/O
+     operations, that exchange a single byte. The intent it to test a reasonable
+     set of blocking and non-blocking scenarios.
+
+     The individual test method names follow their test scenario.
+        [BAccep|SELNBAccep|SPINNBAccep] - Accept either:
+                         blocking, select-non-blocking, spinning-non-blocking
+        [BConn|NBConn] - blocking connect / non-blocking connect
+        [BIO|NBIO]     - blocking / non-blocking I/O operations (read/write)
+        [WR|RW] - connecting thread write/accepting thread reads first, and vice-versa
+        [Id]    - unique test Id
+
+        BAccep_BConn_BIO_WR_1
+        BAccep_BConn_BIO_RW_2
+        SELNBAccep_BConn_BIO_WR_3
+        SELNBAccep_BConn_BIO_RW_4
+        SPINNBAccep_BConn_BIO_WR_5
+        SPINNBAccep_BConn_BIO_RW_6
+        BAccep_NBConn_BIO_WR_7
+        BAccep_NBConn_BIO_RW_8
+        SELNBAccep_NBConn_BIO_WR_9
+        SELNBAccep_NBConn_BIO_RW_10
+        SPINNBAccep_NBConn_BIO_WR_11
+        SPINNBAccep_NBConn_BIO_RW_12
+
+        BAccep_BConn_NBIO_WR_1a         // Non-Blocking I/O
+        BAccep_BConn_NBIO_RW_2a
+        SELNBAccep_BConn_NBIO_WR_3a
+        SELNBAccep_BConn_NBIO_RW_4a
+        SPINNBAccep_BConn_NBIO_WR_5a
+        SPINNBAccep_BConn_NBIO_RW_6a
+        BAccep_NBConn_NBIO_WR_7a
+        BAccep_NBConn_NBIO_RW_8a
+        SELNBAccep_NBConn_NBIO_WR_9a
+        SELNBAccep_NBConn_NBIO_RW_10a
+        SPINBAccep_NBConn_NBIO_WR_11a
+        SPINBAccep_NBConn_NBIO_RW_12a
+    */
```

Comment:
I recognise this test ;-)  I thought there was a version for IP-specific channels, but cannot find it now. I was going to ask if these could be merged or abstracted out somehow.

Reply:
Chris,
Thanks for the comments. I will incorporate them all into the next revision. As regards the IOExchanges test, it uses a @DataProvider of ProtocolFamily with values "UNIX" and "INET" which means the tests are run for TCP sockets. Is that what you were asking? I'm not sure I get the question exactly.
Michael.